pkg/networkservice/common/capturecontext/context.go

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package capturecontext provides a chain element that adds into context reference to current context for passing it
+// from the adapter server/client to the next client/server to avoid the problem with losing values from adapted server/client context.
+package capturecontext
+
+import (
+	"context"
+)
+
+type contextKeyType string
+
+const capturedContext contextKeyType = "capturedContext"
+
+// WithCapturedContext - adds record with nil reference into context or nullifies it
+func WithCapturedContext(ctx context.Context) context.Context {
+	if v := ctx.Value(capturedContext); v != nil {
+		if b, ok := v.(*context.Context); ok {
+			*b = nil
+			return ctx
+		}
+	}
+	var d context.Context = nil
+	return context.WithValue(ctx, capturedContext, &d)
+}
+
+// CapturedContext - returns previously written context
+func CapturedContext(ctx context.Context) context.Context {
+	if val, ok := ctx.Value(capturedContext).(*context.Context); ok && *val != nil {
+		return *val
```

Comment:
I think we need to return ctx in case it is nil value, so we could safely use this method like 

``` ctx = capturecontext.CapturedContext(ctx)``` and check for equality if required to check.

Reply:
ok, fixed

Reply:
It's equally comfortable to return nill here. It allows easy check if context were record.